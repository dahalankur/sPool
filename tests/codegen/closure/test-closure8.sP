int x = 10
shared int y = 10
# (int->int) countup = lambda quack (int num): 
#                         while (num > 0):

#                         ;
#                         return
#                     ;

def quack countup (int num):
    if (num > 0):
        countup(num - 1)
        println(int_to_string(num))
    ;
    return
;

def quack countdownToNum(int num):
    if (y > 0):
        println(int_to_string(y))
        y = y - 1
        countdownToNum(num)
    ;
    return
;

def quack closure((int->quack) down):
    println("counting down")
    down(0)

    println("")
    println("counting up")
    countup(x)
    return
;

print("-----Value of y before: ")
println(int_to_string(y))
closure(countdownToNum)
print("-----Value of y after: ")
println(int_to_string(y))

