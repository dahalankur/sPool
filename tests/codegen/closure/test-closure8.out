-----Value of y before: 10
counting down
10
9
8
7
6
5
4
3
2
1

counting up
1
2
3
4
5
6
7
8
9
10
-----Value of y after: 0
